### PR TITLE
Metrics for scatter queries

### PR DESCRIPTION
**Pull request by tirsen**
We use this to measure impact of introducing lookup tables.

### CODE REVIEW

File: go/vt/vtgate/engine/route.go [hunk 1]
```diff
@@ -152,6 +153,32 @@ var routeName = map[RouteOpcode]string{
 	SelectDBA:         "SelectDBA",
 }
 
+var routeMetrics = map[RouteOpcode]*stats.Counter{
```
Comment (sougou) [inline go/vt/vtgate/engine/route.go]: Use `stats.CountersWithSingleLabel` instead. Example: https://github.com/vitessio/vitess/blob/master/go/vt/vttablet/tabletserver/tabletenv/tabletenv.go#L47.

File: go/vt/vtgate/engine/route.go [hunk 2]
```diff
@@ -152,6 +153,32 @@ var routeName = map[RouteOpcode]string{
...
+	SelectEqual:       stats.NewCounter("RouteSelectEqual", "Counts of statements routed routed to one or more shards using a non-unique vindex. Requires: A Vindex, and a single Value."),
+	SelectIN:          stats.NewCounter("RouteSelectIN", "Counts of statements routed to one or more shards using a Vindex. Requires: A Vindex, and a Values list."),
+	SelectScatter:     stats.NewCounter("RouteSelectScatter", "Counts of statements routed with a scatter query to all shards of a keyspace"),
+	SelectNext:        stats.NewCounter("RouteSelectNext", "Counts of fetches from a sequence"),
+	SelectDBA:         stats.NewCounter("RouteSelectDBA", "Counts of a DBA statements"),
+}
+
+var (
+	scatterWidth = stats.NewCounter("RouteTotalScatterWidth", "Total width (shards routed to by a statement) of all statements routed")
+	queryCount   = stats.NewCounter("RouteQueryCount", "Total statements routed")
```
Comment (sougou) [inline go/vt/vtgate/engine/route.go]: This shouldn't be needed. The total routeMetrics should give you this number.

File: go/vt/vtgate/engine/route.go [hunk 3]
```diff
@@ -152,6 +153,32 @@ var routeName = map[RouteOpcode]string{
...
+)
+
+func init() {
+	stats.NewGaugeFunc("RoutesAverageScatterWidth", "Average width (shards routed to by a statement) of scatters routed", func() int64 {
```
Comment (sougou) [inline go/vt/vtgate/engine/route.go]: This should also not be needed because it's trivially calculated by monitoring tools.

**Comment by demmer** [comment]
I like the intent of this change a lot, but I wonder if it could be made even more generic to cover all statements and not just route.go

It seems to me if we had two counters that both use the plan type as a label, one which counts the queries run at vtgate and one which counts the queries sent to the tablet.

Something like:

```
QueriesProcessed[PlanType]
QueriesRouted[PlanType]
```

(Names could be made better)

Then this would also apply to DMLs and shard-targeted queries.

**Comment by tirsen** [comment]
That's a great idea @demmer!

Unfortunately I don't understand all those other code paths quite as well. Where can I find a conclusive list of "plan types"? Where are all those other places where I need to bump those metrics?

What's the difference between `QueriesProcessed` and `QueriesRouted`?

Since this has already been merged should we rename the metric to something more generic before people start relying on it and then follow up with all the "plan types" and bumping the metric in all those other code paths?

**Pull request merged**